Optimize Random{NumberGenerator}.GetItems for power-of-two choices

stephentoub opened the pull request:
When possible, make a single call to get the required randomness rather than one per element.

Closes https://github.com/dotnet/runtime/issues/92130

```C#
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Running;
using System.Security.Cryptography;

BenchmarkSwitcher.FromAssembly(typeof(Tests).Assembly).Run(args);

[HideColumns("Error", "StdDev", "Median", "RatioSD")]
public class Tests
{
    private const int Length = 11;
    private static readonly string s_charChoices = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789_-";
    private static ReadOnlySpan<int> s_intChoices => [42, 43, 44, 45, 46, 47, 48, 49];
    private static int[] s_dest = new int[Length];

    [Benchmark]
    public string RngGetString() => RandomNumberGenerator.GetString(s_charChoices, Length);

    [Benchmark]
    public void RandGetItems() => Random.Shared.GetItems(s_intChoices, s_dest);
}
```

| Method       | Toolchain         | Mean      | Ratio |
|------------- |------------------ |----------:|------:|
| RngGetString | \main\corerun.exe | 764.18 ns |  1.00 |
| RngGetString | \pr\corerun.exe   |  98.25 ns |  0.13 |
|              |                   |           |       |
| RandGetItems | \main\corerun.exe |  57.21 ns |  1.00 |
| RandGetItems | \pr\corerun.exe   |  16.56 ns |  0.29 |

stephentoub committed:
Optimize Random{NumberGenerator}.GetItems for power-of-two choices

When possible, make a single call to get the required randomness rather than one per element.

jkotas commented on src/libraries/System.Private.CoreLib/src/System/Random.cs:
```diff
@@ -194,6 +196,38 @@ public void GetItems<T>(ReadOnlySpan<T> choices, Span<T> destination)
+            //   and then walk backwards through the destination, filling in the corresponding choice
+            //   based on the random byte at that position when interpreted as bytes. By walking backwards,
+            //   we won't overrite any random bytes that we still need to read.
+            if (!RuntimeHelpers.IsReferenceOrContainsReferences<T>() && Unsafe.SizeOf<T>() >= 2 &&
```
```suggestion
            if (!RuntimeHelpers.IsReferenceOrContainsReferences<T>() && Unsafe.SizeOf<T>() >= 1 &&
```

jkotas commented on src/libraries/System.Private.CoreLib/src/System/Random.cs:
```diff
@@ -194,6 +196,38 @@ public void GetItems<T>(ReadOnlySpan<T> choices, Span<T> destination)
+                BitOperations.IsPow2(choices.Length) && choices.Length <= 256)
+            {
+                // Reinterpret the bottom destination.Length bytes of the destination as our
+                // scratch buffer for randomness.
```
We could generate a block of random bytes to a stack scratch buffer and fill the output one block at a time.

stephentoub replied:
> We could generate a block of random bytes to a stack scratch buffer and fill the output one block at a time.

We can, it just means we either further limit where this can be applied to length stackalloc'able, or we pay the cost of ArrayPool for larger sizes.

jkotas replied:
I meant something like this:
```csharp
Span<byte> randomBytes = stackalloc byte[512];
while (destination.Length > 0)
{
    if (destination.Length < randomBytes.Length)
    {
        randomBytes = randomBytes.Slice(0, destination.Length);
    }

    NextBytes(randomBytes);

    int mask = choices.Length - 1;
    for (int i = 0; i < randomBytes.Length; i++)
    {
        destination[i] = choices[randomBytes[i] & mask];
    }

    destination = destination.Slice(randomBytes.Length);
}
```

stephentoub replied:
> I meant something like this:

I see, trading off multiple Fill/NextByte calls for a temporary array.

bartonjs replied:
Yeah, when I did the AsBytes for GetString (in a comment in the reporting issue) I thought it was OK because no one other than the thread filling that string knew about the target memory.  In a chat I had with @vcsjones on Friday I said (or was convinced so hard by Kevin that I now recall having said it) that I wouldn't like that for GetItemsCore, but it was OK for GetString.

@jkotas 's approach here is more or less what I was thinking when I paged this out of active thinking on Friday.

stephentoub committed:
Address PR feedback

stephentoub committed:
Address PR feedback

Use separate scratch buffer instead of destination buffer. Upside, we don't need to be concerned about temporarily corrupt Ts, and it can work for managed Ts. Downside, it requires separate scratch space, which means an ArrayPool rental for larger sizes.

vcsjones commented on src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/RandomNumberGenerator.cs:
```diff
@@ -350,6 +353,38 @@ private static void GetHexStringCore(Span<char> destination, bool lowercase)
+            //   a scratch buffer. we can use a single RNG.Fill to get all of the required randomness,
+            //   and then walk backwards through the destination, filling in the corresponding choice
+            //   based on the random byte at that position when interpreted as bytes. By walking backwards,
+            //   we won't overrite any random bytes that we still need to read.
```
```suggestion
            //   we won't overwrite any random bytes that we still need to read.
```

vcsjones commented on src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/RandomNumberGenerator.cs:
```diff
@@ -350,6 +353,44 @@ private static void GetHexStringCore(Span<char> destination, bool lowercase)
+                    (arrayPoolArray = ArrayPool<byte>.Shared.Rent(destination.Length));
+                randomBytes = randomBytes.Slice(0, destination.Length);
+
+                Fill(randomBytes);
```
It will probably get inlined, but we can go directly to the internal implementation here.

```suggestion
                RandomNumberGeneratorImplementation.FillSpan(randomBytes);
```

vcsjones commented on src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/RandomNumberGenerator.cs:
```diff
@@ -350,6 +353,44 @@ private static void GetHexStringCore(Span<char> destination, bool lowercase)
+
+                // Get a scratch buffer to fill with random bytes, either stack space if the
+                // destination is small enough, or an array pool array if it's larger.
+                // If T is unmanaged, we _could_ use the destination buffer itself, reinterpreted
```
`T` is `char` is probably the most interesting case to optimize. I would be in favor of keeping the re-interpret for special casing `char` since random strings are the most interesting case and the impetus for this API.

If we feel that adds too much complexity for a narrow case of "input len is pow2" then that's understandable too.

vcsjones commented on src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/RandomNumberGenerator.cs:
```diff
@@ -350,6 +353,44 @@ private static void GetHexStringCore(Span<char> destination, bool lowercase)
+                byte[]? arrayPoolArray = null;
+                Span<byte> randomBytes = destination.Length <= StackAllocThreshold ?
+                    stackalloc byte[StackAllocThreshold] :
+                    (arrayPoolArray = ArrayPool<byte>.Shared.Rent(destination.Length));
```
```suggestion
                    (arrayPoolArray = CryptoPool.Rent(destination.Length));
```

stephentoub committed:
Address PR feedback, again

Loop over a stackalloc'd buffer rather than using ArrayPool

bartonjs commented on src/libraries/System.Private.CoreLib/src/System/Random.cs:
```diff
@@ -194,6 +197,38 @@ public void GetItems<T>(ReadOnlySpan<T> choices, Span<T> destination)
+            // In such a case, we can grab a bunch of random bytes in one call.
+            if (BitOperations.IsPow2(choices.Length) && choices.Length <= 256)
+            {
+                Span<byte> randomBytes = stackalloc byte[512]; // arbitrary size, a balance between stack consumed and number of random calls required
```
The line length with this comment post-statement feels a bit long.

```suggestion
                // This is an arbitrary size, a balance between stack consumed and number of random calls required
                Span<byte> randomBytes = stackalloc byte[512];
```

looks happier (more in line with the rest of the file (or, at least, PR))

Ponant commented:
I few remarks if you do not mind.

1) I tested @bartonjs method against mine and confirm it is faster, but it is mainly due to the use bitwise AND instead of integer division, and not theunsafe territory. At least on my machine.

2) I think the benefit of the pow2 non-negligible speedup should be documented in th emethods that use it such as GetString, perhaps...

3) I have an addional suggestion, which I think can be beneficial to users and which can be implemented in the same PR. But I do not know if oyu prefer that I create another issue or drop it here? it is about rewritting GetBytes to allow users to have a byte arrays where each byte is bound to [0,N] where N is a power of two. For N=256, we recover the case of the current getBytes. Let me know I will drop a code

stephentoub commented:
> I have an addional suggestion, which I think can be beneficial to users and which can be implemented in the same PR. But I do not know if oyu prefer that I create another issue or drop it here? it is about rewritting GetBytes to allow users to have a byte arrays where each byte is bound to [0,N] where N is a power of two. For N=256, we recover the case of the current getBytes. Let me know I will drop a code

This sounds very niche, and the savings is effectively an array access per element. I don't think it's worth special-casing without seeing more examples of where it would be really meaningful.

> I think the benefit of the pow2 non-negligible speedup should be documented in the methods that use it such as GetString, perhaps...

It's an implementation detail. We avoid documenting such optimization implementations, as we reserve the right to change them at any point.

Ponant commented:
> > I have an addional suggestion, which I think can be beneficial to users and which can be implemented in the same PR. But I do not know if oyu prefer that I create another issue or drop it here? it is about rewritting GetBytes to allow users to have a byte arrays where each byte is bound to [0,N] where N is a power of two. For N=256, we recover the case of the current getBytes. Let me know I will drop a code
> 
> This sounds very niche, and the savings is effectively an array access per element. I don't think it's worth special-casing without seeing more examples of where it would be really meaningful.
> 

A) Generate random booleans
B) Generate Random vectors in 4 dimensions (my case)

stephentoub commented:
> > > I have an addional suggestion, which I think can be beneficial to users and which can be implemented in the same PR. But I do not know if oyu prefer that I create another issue or drop it here? it is about rewritting GetBytes to allow users to have a byte arrays where each byte is bound to [0,N] where N is a power of two. For N=256, we recover the case of the current getBytes. Let me know I will drop a code
> > 
> > 
> > This sounds very niche, and the savings is effectively an array access per element. I don't think it's worth special-casing without seeing more examples of where it would be really meaningful.
> 
> A) Generate random booleans B) Generate Random vectors in 4 dimensions (my case)

I don't follow.  Are you proposing a new API? Or are you proposing that the implementation special-case `typeof(T) == typeof(bool)` and validates that `choices[0] == false` and `choices[1] == Unsafe.As<byte, bool>(1)`?  And for `typeof(T) == typeof(byte)`, looping through all of the choices as a precompute step to make sure that `choices[i] == (byte)i`?

GrabYourPitchforks commented:
@stephentoub Consider also duplicating the test here with one that passes a power-of-two sized values span. That way we get distribution tests across both implementations which now exist in this method.

https://github.com/dotnet/runtime/blob/0ad4e699b0a6d92c2d08f7b9b6393448ab8f310a/src/libraries/System.Security.Cryptography/tests/RandomNumberGeneratorTests.cs#L669-L675

Ponant commented:
The current `GetBytes` returns and array of bytes, where each byte can take any value from 0 to 255. My proposal is to have GetBytes to also allow me to return an array of bytes but where each byte is bound to [0,N] where N is a powere of two. For example, I can generate a byte array of bools  011100011010101 or random quaternions or any 4d vector, or octonions or above powers.

I am using something like that, it is essentially the same spirit as you do here with the new GetString in this PR, but for getting byte arrays. I am using something like that (checks on parameters omitted for brievety)

```
private static byte[] GetRandomBytes(int length, int maxNumOfBytes)
 {
        Span<byte> bytes = length <= _stackAllocThreshold ? stackalloc byte[length] : new byte[length];

        // Generate random bytes, each taking values between 0 and 255
        RandomNumberGenerator.Fill(bytes);

        // We do not need to loop if we cover all byte possibilities, since  RandomNumberGenerator.Fill does this 
        if (maxNumOfBytes != 256)
        {
            byte mask = (byte)(maxNumOfBytes - 1);

            // make each byte between 0 and maxNumOfBytes
            for (var i = 0; i < bytes.Length; i++)
                bytes[i] &= mask;

        }

        return bytes.ToArray();
}
```

stephentoub commented:
> The current GetBytes returns an array of bytes, where each byte can take any value from 0 to 255. My proposal is to have GetBytes to also allow me to return an array of bytes but where each byte is bound to [0,N] where N is a power of two.

That's a new API.  Feel free to open a separate proposal for one.  That won't be done as part of an optimization PR.

vcsjones commented:
> The current GetBytes returns an array of bytes, where each byte can take any value from 0 to 255. My proposal is to have GetBytes to also allow me to return an array of bytes but where each byte is bound to [0,N] where N is a power of two. For example, I can generate a byte array of bools 011100011010101 or random quaternions or any 4d vector, or octonions or above powers.

This would be a new [API proposal](https://github.com/dotnet/runtime/blob/main/docs/project/api-review-process.md).

We can discuss the merits and usefulness of the proposal there to keep this PR focused on the perf of existing APIs.

stephentoub committed:
Merge branch 'main' into randpow2

stephentoub committed:
Tweak test to validate variable number of choices

stephentoub commented:
@bartonjs, per your feedback on my recent PR, I've marked this as a breaking change to be documented.

stephentoub commented:
Opened https://github.com/dotnet/runtime/pull/108018 to revert the breaking change aspect.
